active/0024-message-validation.md

```diff
@@ -0,0 +1,132 @@
+
+## Motivation
+
+### Message Validation
+
+Technically, by using EMQX rules engine, it is possible to validate the incoming
+MQTT messages before sending the message off to data bridges.
+
+For instance, one can write a rules engine SQL like below to
+drop messages which are not avro encoded and also when the 'val' field is not greater than 10.
+
+```
+SELECT schema_decode(my_avro_schema, payload) as decoded from t/# where decoded.val > 10
+```
+
+However, currently rules are only used in data bridges, but they do not interfere MQTT message pub/sub.
+For example, if there is client subscribed to `t/#` it would still receive the avro encoded message.
+
+If we also want to stop the message being published, then there would be a need for a message republish action,
+and the publisher and subscriber will likely be forced to use different topics.
+e.g. publisher publishes to `t1/...`, republish action send it to `t2/...` for subscribers to receive.
+
+This can work, but not easy to use.
+
+### Message filter
+
+Message filtering can be achieved by confugring message validation rule to drop invalid messages.
+
+## Design
+
+This data validation step should happen after authorization and before rules engine.
+Probably can make use of the 'message.publish' with a hook priority which
+positions the callback after authorization, but before data bridges.
+
+### High level design
+
+* Make use of the rules engine SQL, but without the `FROM` clause as it's always the MQTT message being th input.
+
+Each data validation rule should have a unique ID. And each rule should have a input topic-filter name which
+can be used to build a topic index for quick look-up.
+
+The common parts can be described as hocon config below:
+
+### Message Validation
+
+```
+validation1 {
+    enable = true # or 'false'
+    type = validation
+    description = "drop message if it is not compabitble with my avro schema or if payload.value is less tha 0"
+    topics = "t/#" # or topics = ["t/1", "t/2"]
+    check_mode = any_pass # or all_pass
+    failure_action = disconnect # (disconnect also implies 'dorp') or 'drop' to only drop the message
+    checks = [
+        {
+            # message payload payload to be validated against JSON schema
+            type = json # can also be 'avro' or 'protobuf'
+            schema = "my-json-schema-registration-id"
+        }
+        {
+            # SQL evaluates to empty result indicates validation falure
+            # in this example, if payload.value is less than 0, the message is considiered invalid
+            type = sql
+            sql = "SELECT * WHERE payload.value < 0"
+        }
+    ]
+}
+```
+
+If there are more than one validation matched for one message, all validations should be executed.
+For example, if one is configured with `topics = "t/#"` and another with `topics = "t/1"`,
+when a message is published to `t/1`, the both validations should be triggered (without ordering guarantee).
```

Comment:
Few more random thoughts (sorry).

If we are to stick to more or less the same concept of rules + actions, we may think about introducing some kind of _ingress rules_ instead, with a slightly different scope: filtering / transformation of incoming messages _before_ they hit the broker / routing / hooks. One possible action at this point is just "publish", possibly with a few knobs to tweak. For example, with rules like `SELECT 'dlq/' + topic as topic, avro_decode('schema_id', payload) as payload FROM t/# WHERE clientid like 'FW%'` we could provide both filtering, transformation and selective topic rewriting.

Similarly to the current design, neither ordering nor special handling of overlapping topic filters should be needed (supposedly). If no rules matched, the default would be to just publish message as is. I'm wondering if this could also cover the use cases for the Message Transformation EIP.

The need to disconnect clients though doesn't fit very well here I must admit, I can think of:
* Adding a knob to this "publish" action that says what to do if there's nothing to publish.
* Writing a rule like `SELECT clientid FROM '$events/ingress/message.dropped'` and attaching a "disconnect" action to it.
* Similar to the one above, but within the authz realm.

Reply:
The journey of my iterations for this feature (partially reflected in this EIP commit history) can be summarized as a progressive shift from "Programing in SQL" towards "Programing in JSON".
I am still not quite sure if this current state is a striking a perfect balance in between, so thank you for your thoughts, it should help to stop me from drifting to far away from the SQL.

> Similarly to the current design, neither ordering nor special handling of overlapping topic filters should be needed (supposedly). If no rules matched, the default would be to just publish message as is. I'm wondering if this could also cover the use cases for the Message Transformation EIP.

Now we have to discuss this: if a rule outputs a new topic which matches  another rule ?


> Similar to the one above, but within the authz realm.
do you mean something like
`authorization.validation_on_failure = disconnect | drop` ?
this would imply that the config is globally applied to all clients, all topics and all messages ?

Reply:
> a progressive shift from "Programing in SQL" towards "Programing in JSON".

Understood. I see the point. I agree that it might overall be better UX-wise. Maybe I'm attaching too much importance to the goal of sticking to the same concepts familiar to the customers (though, arguably, they may cover more use cases rather than less). Also it seems inevitable that each new "programmable in JSON" thing will bring a set of its own concepts, slightly different from any other, which may breed complexity.

> Now we have to discuss this: if a rule outputs a new topic which matches another rule ?

Very good question. I would either lean towards the safe side: do nothing special. Or even disallow this, it was sort of an illustration of potential expressiveness anyway. Curious: what happens now in a similar situation with topic rewrite rules?

> do you mean something like
> this would imply that the config is globally applied to all clients, all topics and all messages ?

I have to admit that I don't understand the use case well enough, so that's more like a speculation. Might be something like a separate authorizer in the chain with:
```
SELECT try(avro_schema_decode(schema_id, payload)) as result, 'disconnect' as action WHERE result.error IS NOT NULL
FROM 't/topic/filter/with/avro/payloads/#'
```
Where the default action is `ignore` (i.e. when nothing was produced by the SQL expression).

---

Tangentially: disconnecting misbehaving clients sounds questionable from the resource saving perspective: it's not clear that handling consecutive reconnects is cheaper than just dropping messages.

Reply:
> Curious: what happens now in a similar situation with topic rewrite rules?

Matched only once.

> 'disconnect' as action 

Thanks, then I know it's not a global config, but a even more fine-grained control flow.